### PR TITLE
Implement memory address folding into Intel HW instrinsic memory load instruction address operand (1/2)

### PR DESCRIPTION
**Pull request by 4creators**
It enables folding of pointer based addressing. Folding of struct/class member addresses requires additional work. 

Fixes https://github.com/dotnet/coreclr/issues/19521

Cases fixed are similar in form to:

```C#
float *address = &matrix;
var row1 = Sse.LoadVector128(address);
var row2 = Sse.LoadVector128(address+4);
var row3 = Sse.LoadVector128(address+8);
var row4 = Sse.LoadVector128(address+12);
```

Disasm for above code fragment is as follows:
```asm
IN0007: 00005D lea      rcx, bword ptr [V00 rsp+20H]
IN0008: 000062 vmovups  xmm6, xmmword ptr [rcx]
IN0009: 000067 vmovups  xmm7, xmmword ptr [rcx+16]
IN000a: 00006D vmovups  xmm8, xmmword ptr [rcx+32]
IN000b: 000076 vmovups  xmm9, xmmword ptr [rcx+48]
```

### CODE REVIEW

File: src/jit/lowerxarch.cpp [hunk 1]
```diff
@@ -2521,42 +2533,83 @@ bool Lowering::IsContainableHWIntrinsicOp(GenTreeHWIntrinsic* containingNode, Ge
     noway_assert(supportsRegOptional != nullptr);
     *supportsRegOptional = supportsGeneralLoads;
 
-    if (!node->OperIsHWIntrinsic())
+    if (!node->OperIsHWIntrinsic() && !(category == HW_Category_MemoryLoad && nodeOper == GT_LEA && supportsContainment))
```
Comment (4creators) [inline src/jit/lowerxarch.cpp]: This temporary solution until whole category will have support for address folding

**Comment by tannergooding** [comment]
I believe the discussions, so far, were that we wanted to eventually remove the special handling for the Load/Store intrinsics and merge it back with the existing Indir node handling (with some mechanism to carry the additional info, such as aligned or non-temporal).
CC. @CarolEidt, @fiigii 

For reference: https://github.com/dotnet/coreclr/issues/18407

**Comment by CarolEidt** [comment]
@tannergooding - yes, I still think that's the direction we want to pursue, though as @mikedn pointed out, my original thought that we'd put the attributes on the `GenTreeAddrMode` may be probably the wrong design direction, as it makes the semantics of the indirection dependent upon its children. But IIRC the `GTF_IND` flags are already fully subscribed, so we'd either need to make room for more of them, or determine if we can fit an additional flags field in the `GenTreeIndir` node. Note that the block nodes derive from that, so we have to take them into account as well. I'm not sure whether we'd ever support the additional attributes on block nodes, but we need to at a minimum ensure that we don't break their size requirements.

**Comment by mikedn** [comment]
> But IIRC the GTF_IND flags are already fully subscribed, so we'd either need to make room for more of them, or determine if we can fit an additional flags field in the GenTreeIndir node

Well, for containment purposes we don't care about non-temporal indirs so we should leave them alone. So what's left (in the useful sense) are just `Load` and `LoadAligned`. Perhaps we can reuse the `GTF_IND_UNALIGNED` somehow.

**Comment by mikedn** [comment]
> so we'd either need to make room for more of them

There may be a spare bit - `GTF_IND_ARR_LEN`. It's used but it appears to be bogus - set by the importer in minopts/debug mode and tested by VN which is not supposed to run in minopts/debug mode.

If we get rid of `GT_ASG` we also won't need `GTF_IND_ASG_LHS` anymore but that's going to take a while probably.

**Comment by 4creators** [comment]
This PR essentially remaps the GT_IND processing path and all used mechanism for pointer arguments to HW intrinsics processing path. If we do not want to pursue that direction I will close PR. 

However, it is quite useful at assessing performance gains from a bit better load handling in my tests.

**Comment by tannergooding** [comment]
> set by the importer in minopts/debug mode and tested by VN which is not supposed to run in minopts/debug mode

Wouldn't this get hit in `Checked` mode, where we do run some optimizations and also do all the asserts/etc?

**Comment by 4creators** [comment]
Although, after reading through several discussions and looking back at this work I would prefer not to use indir node. The most important argument is that it makes any IR tree transformations hard and risky without multiple additional checks. It was main reason I have not implemented some indir paths with this PR. And citing @mikedn 

> This means that various IR transforms need to be careful not to break the tree

I would rather go for more clear cut IR semantics (not that my opinion is very important since I am not compiler engineer 😄 ).

**Comment by mikedn** [comment]
> This means that various IR transforms need to be careful not to break the tree

That was referring to having an extra node to hold additional information about the indir. Using `GT_IND` for "normal" SIMD loads such as `Load`/`LoadAligned` would be the normal thing to do in the IR, if it weren't for this alignment flag problem. And possible base type info, though that can probably be ignored for now.

> Wouldn't this get hit in Checked mode, where we do run some optimizations and also do all the asserts/etc?

I'm talking about JIT compilation modes. I suspect that you're talking about JIT build configurations.

**Comment by 4creators** [comment]
> Using GT_IND for "normal" SIMD loads such as Load/LoadAligned would be the normal thing to do

The problem here seems that this node cannot hold enough information about intrinsics which are going to be emitted.

One of the solution could be using GT_IND attached as obligatory address operand of GT_HWintrinsic node. Than implementation could be created by deriving from `GenTreeHWIntrinsic` and adding restrictions via methods on type of address operands accepted.

**Comment by mikedn** [comment]
> The problem here seems that this node cannot hold enough information about intrinsics which are going to be emitted.

Eh, it's doable. It's just nobody actually tried. And to be clear, I'm only talking about `Load`/`LoadAligned`, I do not think it's reasonable or useful to attempt to somehow fit other complex load/store intrinsics in `GT_IND`.

> One of the solution could be using GT_IND attached as obligatory address operand of GT_HWintrinsic node.

I'm not sure what you mean by that but it seems like you're suggesting to use a `GT_IND` with a different meaning that the current one. This just doesn't work well in a proper IR. And JIT's IR already has such problems that adding more is IMO out of the question. In fact `GT_IND` itself already suffers from this - it's either a load or a stored, depending on where in the tree it appears.

**Comment by mikedn** [comment]
All this got me wondering if using `GT_IND` actually helps with containment. You'll probably still need to special case the existing containment code due to alignment issues. VEX instructions can't contain aligned indirs and non-VEX instructions can't contained unaligned indirs. Hrm.

**Comment by 4creators** [comment]
>  I'm only talking about Load/LoadAligned

It would be OK but as you said indir is everything at the same time: represents both load and stores. Adding more would probably make things even worse.

> I'm not sure what you mean by that but it seems like you're suggesting to use a GT_IND with a different meaning

The proposal is to use "decorator" pattern by changing type of op1 or op2 which hold address operand to GT_IND.

**Comment by 4creators** [comment]
Actually we are hitting several issues due to implementation of IR. Would it be possible to move to SSA only form and change IR nodes implementation without API changes to improve  overall implementation flexibility? Just to expand information content available in nodes and have more flexibility in code transformations.

**Comment by mikedn** [comment]
> The proposal is to use "decorator" pattern by changing type of op1 or op2 which hold address operand to GT_IND.

`GT_IND` is supposed to take an address and return whatever value is stored there. You seem to suggest treating it as some kind of "address operand", which is not.

The "decorator" pattern does work, but not in this form. What you need is a node "extension" - an object that can be optionally attached to a tree node to hold extra information that is not commonly needed. But that object is not a tree node itself and it's not part of the IR tree/linear order. However, that would probably be overkill in this case. And it too needs space in the tree node, you need a pointer or something to associate that extension object with the tree node.

**Comment by mikedn** [comment]
> Actually we are hitting several issues due to implementation of IR.

Eh, IMO the JIT needs an IR overhaul more than anything else. e.g. #19412 is one really nasty issue and there a few more that would be definitely worth fixing.

> Would it be possible to move to SSA only form and change IR nodes implementation without API changes to improve overall implementation flexibility?

I'm not sure what exactly you mean by SSA only form. But in any case, issues like `GT_ASG` needs to be fixed first. They cause problems to the existing SSA implementation as well.

**Comment by CarolEidt** [comment]
> I would rather go for more clear cut IR semantics

Having non-indir nodes that perform indirections is *not* clear-cut IR semantics.

> indir is everything at the same time: represents both load and stores. Adding more would probably make things even worse.

I don't think so; and the issue of representing both loads and stores is not a problem in LIR (unless I misunderstand what's being said), and ideally will go away as we rationalize the front-end IR (and getting rid of `GT_ASG` is the most important rationalization).

Moving toward an IR that more faithfully and intrinsically represents SSA is a great goal; but not one that can be achieved in the near term.

Rather than a "decorator" or side data structure (which is anathema to me), we could consider a node derived from `GenTreeIndir`, with a distinguished operator such as `GT_IND_INTRINSIC`.

**Comment by fiigii** [comment]
> One of the solution could be using GT_IND attached as obligatory address operand of GT_HWintrinsic node. 
>op1 or op2 which hold address operand to GT_IND.

@4creators It would be problematic and very odd if use a load/store (`GT_IND `) as other load/store(`GT_HWintrinsic `)'s operand to hold the address. I guess what you want is something like @CarolEidt's previous suggestions (in other threads) that encodes HWintrinsic info into address mode nodes. 

> we could consider a node derived from GenTreeIndir, with a distinguished operator such as GT_IND_INTRINSIC.

I like this solution instead of any other workaround.

**Comment by 4creators** [comment]
> Rather than a "decorator" or side data structure (which is anathema to me), we could consider a node derived from GenTreeIndir, with a distinguished operator such as GT_IND_INTRINSIC.

Checked how this would work and it seems that `GenTreeIndirIntrinsic` derived from `GenTreeIndir` would solve the problem as it has enough space to hold all necessary info for HW intrinsics. `GenTreeIndir` node size is exactly the same as the `GenTreeOp` base of `GenTreeJitIntrinsic`. It means that it has enough space for all HW and SIMD intrinsic specific data.

**Comment by CarolEidt** [comment]
>> Having non-indir nodes that perform indirections is not clear-cut IR semantics.

> Here I have to disagree with you. Any node that performs a clearly defined operation and it always performs that operation regardless of context is clear-cut IR semantics :)

I don't disagree with this, but I think that a situation where `OperIsIndir()` either a) would return false for a node that *does* do an indirection, or b) would return true, but wouldn't then allow me to use `AsIndir()` would not be clear-cut IR semantics. This is already an issue with the atomic ops, and is why we also need `OperIsStore()` but I would prefer to avoid complicating this even further.

**Comment by mikedn** [comment]
>  but I think that a situation where OperIsIndir() either a) would return false for a node that does do an indirection, or b) would return true, but wouldn't then allow me to use AsIndir() would not be clear-cut IR semantics.

Hmm, I see where you're coming from but I'm not sure this is a problem. Or to be more precise, if this is THE problem. For one thing, we could make `OperIsIndir` recognize hw intrinsics that are indirections. Ah, right, `AsIndir` will then fail because the node class is different. 

That's a different problem - that even today there's no clear correlation between `AsX` and the zillion of `OperIsX`. The former have a precise meaning and are automatically generated from `gtstructs.h`. The later are a mess, added by various people at various times for various purposes. If you want this fixed then it's probably easy - use the existing `gtstructs.h` to generate `IsX` functions such that when `IsX` returns true you know that `AsX` also succeeds. But that has less to do with the indir problem.

Back to indirs - it's not clear what actual problem the "extended" version of `OperIsIndir` would solve. HW intrinsics still have special properties that may require special handling here and there. And adding a derived class means that some code will needs to specifically check for that and do whatever needs to be done. In short, you'll have some special casing one way or another.

Even just using `GT_IND` and a flag for `LoadAligned` may require some special casing - so far indirections had a `GTF_EXCEPT` side effect due to the address potentially being null. If you can prove that the address is never null then the side effect can be ignored. With `LoadAligned` you can get an exception if the address is not aligned so proving that is not null is no longer sufficient to ignore the side effect. 

Well, assuming that anyone cares about such a side effect. But the point is that all this back and forth, new opers, new node classes and whatnot does not necessarily avoid the need for special casing.

Something that might work is a different level of abstraction - instead of `OperIsIndir` have `GetIndirInfo` that returns various information such as the address tree, the indirection type/size and side effects. Any code that cares about indirs would just call that and if new kinds of indirs are added you only need to modify that function. Hopefully.

Anyway, I'm not against adding a derived indir class, if it actually works. I'm just not convinced it significantly improves things. What I'm particularly against is adding extra nodes to care extra information.

**Comment by mikedn** [comment]
At least some of the LEA issues I've seen in the past would likely require SSA to be fixed. The existing SSA form, while functional it's incomplete and the SSA graph traversal is not very efficient (in part due to the pesky GT_ASG). So IMO we're back at square one - there CQ improvements that could be done but require an overhaul of the JIT's IR. Of course, that's just my opinion. The MS team may have a different opinion.

But I took a look at that matrix example and it seems that the problem is rather different. Before morph we have something like:
```
               [000051] ---XG-------              /--*  HWIntrinsic simd16 float LoadVector128
               [000050] ----G----U--              |  \--*  CAST      long <- ulong <- byref
               [000049] ----G-------              |     \--*  ADDR      byref 
               [000048] ----G-------              |        \--*  FIELD     float  M41
               [000047] L-----------              |           \--*  ADDR      byref 
               [000046] ------------              |              \--*  LCL_VAR   struct V00 arg0         
               [000054] -A-XG-------              *  ASG       simd16 (copy)
               [000052] D-----------              \--*  LCL_VAR   simd16(AX) V05 tmp4         
```
Looks relatively reasonable. There's a stupid cast in the way (perhaps it's the actual problem, I haven't checked yet) but other than that nothing out of the ordinary. The ADDR/FIELD/ADDR/LCL_VAR chain should be folded to something simpler. But it does not. Morph generates:
```
               [000051] -A-XG+------              /--*  HWIntrinsic simd16 float LoadVector128
               [000085] -----+------              |  |  /--*  LCL_VAR   long   V09 tmp8         
               [000087] -A--G+------              |  \--*  COMMA     long  
               [000088] -----+------              |     |     /--*  CNS_INT   long   48 field offset Fseq[M41]
               [000089] -----+------              |     |  /--*  ADD       byref 
               [000046] -----+------              |     |  |  \--*  LCL_VAR   byref  V00 arg0         
               [000084] -A--G+------              |     \--*  ASG       long  
               [000083] D----+-N----              |        \--*  LCL_VAR   long   V09 tmp8         
               [000054] -A-XG-------              *  ASG       simd16 (copy)
               [000052] D---G+-N----              \--*  LCL_VAR   simd16(AX) V05 tmp4         
```
Scratching head. I actually noticed something like this recently but didn't pay too much attention to it.

**Comment by 4creators** [comment]
Closing as another approach to implementation expected

**Pull request closed without merging**